vcf.c

```diff
@@ -376,6 +376,96 @@ int bcf_hrec_find_key(bcf_hrec_t *hrec, const char *key)
     return -1;
 }
 
+static int bcf_hrec_set_type(bcf_hrec_t *hrec)
+{
+    if ( !strcmp(hrec->key, "contig") ) hrec->type = BCF_HL_CTG;
+    else if ( !strcmp(hrec->key, "INFO") ) hrec->type = BCF_HL_INFO;
+    else if ( !strcmp(hrec->key, "FILTER") ) hrec->type = BCF_HL_FLT;
+    else if ( !strcmp(hrec->key, "FORMAT") ) hrec->type = BCF_HL_FMT;
+    else if ( hrec->nkeys>0 ) hrec->type = BCF_HL_STR;
+    else hrec->type = BCF_HL_GEN;
+    return 0;
+}
+
+/**
+    bcf_hrec_check() - check the validity of structured header lines
+
+    Returns 0 on success or negative value on error.
+
+    Currently the return status is not checked by the caller
+    and only a warning is printed on stderr. This should be improved
+    to propagate the error all the way up to the caller and let it
+    decide what to do: throw an error or proceed anyway.
+ */
+static int bcf_hrec_check(bcf_hrec_t *hrec)
+{
+    int i;
+    static uint8_t valid_ctg[255], valid_tag[255], valid_set = 0;
+    if ( !valid_set )
+    {
+        for (i=0; i<255; i++) valid_ctg[i] = 0;
+        for (i=(int)'0'; i<=(int)'9'; i++) valid_ctg[i] = 1;
+        for (i=(int)'A'; i<=(int)'Z'; i++) valid_ctg[i] = 1;
+        for (i=(int)'a'; i<=(int)'z'; i++) valid_ctg[i] = 1;
+        char valid[] = "!#$%&*+./:;=?@^_|~-";
+        for (i=0; i<sizeof(valid)-1; i++) valid_ctg[(int)valid[i]] = 1;
+
+        for (i=0; i<255; i++) valid_tag[i] = 0;
+        for (i=(int)'0'; i<=(int)'9'; i++) valid_tag[i] = 1;
+        for (i=(int)'A'; i<=(int)'Z'; i++) valid_tag[i] = 1;
+        for (i=(int)'a'; i<=(int)'z'; i++) valid_tag[i] = 1;
+        valid_tag[(int)'_'] = 1;
+        valid_tag[(int)'.'] = 1;
+    }
+
+    bcf_hrec_set_type(hrec);
+
+    if ( hrec->type==BCF_HL_CTG )
+    {
+        i = bcf_hrec_find_key(hrec,"ID");
+        if ( i<0 ) goto err_missing_id;
+        char *val = hrec->vals[i];
+        if ( val[0]=='*' || val[0]=='=' || !valid_ctg[(int)val[0]] ) goto err_invalid_ctg;
+        while ( *(++val) )
+            if ( !valid_ctg[(int)*val] ) goto err_invalid_ctg;
+        return 0;
+    }
+    if ( hrec->type==BCF_HL_INFO )
+    {
+        i = bcf_hrec_find_key(hrec,"ID");
+        if ( i<0 ) goto err_missing_id;
+        char *val = hrec->vals[i];
```

Comment:
This needs to be unsigned char, to avoid underflow of the valid_ctg array.  Similarly all the other lookups to this array.

Note casting it to int may remove the compiler warning, but it won't solve the underlying issue.